Save resampled images into a folder structure indicating transformations

JorisDebonnet opened the pull request:
This is a solution for issue #4295. 

Resampled images are now saved into a folder structure that indicates the image transformations that have happened, rather than creating a long filename prefix. 

This change should be transparent to users, so I'm hoping this may make it into 3.2. I have tracked down all implications I could find, but will be happy to create additional patches if any issues come up during beta testing. 

(Note that `Image::getGeneratedImages` uses the SPL function `RecursiveIteratorIterator`. SPL is included and compiled by default since PHP 5.0, and can no longer be disabled since PHP 5.3.)

JorisDebonnet commented:
(closed and reopened to retrigger a Travis CI build, since one of them timed out... it should be all fine since the problematic one built succesfully before)

jonom commented:
Heck yeah! Awesome work. I tried it out locally and worked great for me. The old resampled images will hang around until manually deleted, not sure if that's an issue or not and the same thing probably happens for images that use deprecated function names too. Could be nice to have a "Delete all resampled" task in the dev task menu that destroys all `_resampled` folders and their contents but that's probably outside the scope of this PR and think that would cause issues for images referenced in HTMLEditorField as I think they're only generated when the field is saved, not when the content is viewed. But I'm rambling... nice work on this @JorisDebonnet, hope this gets accepted.

jonom commented:
Hey @JorisDebonnet actually found one problem here but probably not due to this PR - if you try to resize any images that are already in HTMLEditorField instances they keep referring to the previously generated image using the old filename format and a new one is not generated to suit the new size.

This will probably affect all images going from 3.1 to 3.2 even without this PR because of the encoded filenames I think? Looking at the way HTMLEditorField stores local image URLs I'm surprised to see it's not stored as a short code that is translated later (like internal page links), but instead is a fully evaluated URL to the resampled image. This is kind of a hack and means we're relying on a very specific output format from an API (abusing it). Probably would be a lot cleaner and less prone to breakage if images were stored as short codes and generated when the content is parsed for a template? I'd say this is the root cause of the old bug where flushing generated images for an `Image` object causes that image to become a broken link on all pages where the image is referenced in HTMLText content, requiring you to save the page again to regenerate the images.

@silverstripe/core-team any thoughts on this? I think there may be a need here in 3.2 for images in HTMLEditorField to be made to use short codes and a dev script provided to convert all existing image tags to the short code format which should safeguard against future filename output changes. The alternative would be to revert all the changes to `Image` that have gone in I guess, but I don't think that should be required because the current way of storing image tags isn't really API-safe and could be considered broken...

JorisDebonnet commented:
Thanks Jono! :)

And I agree about the problematic way Images are saved in the `HTMLEditorField`s. I encountered a similar issue months ago already in 3.1, due to the resized images on your pages that stop working when you clear out the `_resampled` directory. Intuitively, you'd expect those images to be automatically generated again, so indeed, a shortcode would be the best way to go. It only needs the resize and scale methods (right?), so shouldn't be *that* complicated. 

Even if we do add a shortcode system in the future though, people will still have issues with existing pages because those still contain the old resized images. To facilitate an easier transition period, we'd need to add some extra code to convert "old resized images" into shortcodes. We could still be able to finish that before the release candidate. But let's create a separate issue for that!

JorisDebonnet commented:
Actually, I'm looking into "recognizing old format images" now. Let's see if I can figure that out and add it to this PR.

JorisDebonnet commented:
Well -- I fixed it. It was in fact something introduced by my PR, and it's now fixed! Meaning that "old format images" are now recognized, and will be saved into the new format when the page is saved again.

jonom commented:
Nice! Well that takes care of that part then I guess :) I've opened a separate issue for the shortcode idea: https://github.com/silverstripe/silverstripe-framework/issues/4337

jonom commented:
Hey @JorisDebonnet just wondering if you've accounted for empty directory cleanup here? Might be a good idea to check for empty folders during Image::deleteFormattedImages(). For each image being deleted I guess you'd check if the containing folder is empty and delete it if so, then keep traversing up doing the same thing until you find a non-empty folder or get to the _resampled folder.

JorisDebonnet commented:
I actually gave some thought to that, but then considered that you usually only need to delete a format-folder when you change a template. Because even if a format-folder becomes empty, a new image that ends up in the same template will undergo the same transformation, and will use that same folder again. 

Of course images may also be formatted without templates (or other hard-coded sections). I guess that only happens when an image in the CMS is resized. So for that particular situation, it is indeed useful to add a check to `Image::deleteFormattedImages`. I personally believe it will only very very rarely trigger a folder removal, because it only happens if you actually *delete* the image (not if you just change the image's size on the page) -- but I suppose that's beside the point. 

So I've added that, including a new `Filesystem::removeFolderIfEmpty` method (it felt wrong to put all that code in `Image.php` and not share it there). One remaining point perhaps is that `Image::deleteFormattedImages()` does not delete formatted images in the "old" format (SetWidth400)... but of course that's the case before this PR, too. I could add code to search for those as well but that's probably going too far.

jonom commented:
:+1: Thanks for that, I think that's a nice addition

JorisDebonnet commented:
Forgive me for 'bumping' -- I'm just curious whether this might get a milestone tag :)

Any remaining concerns about these changes? I realise it may feel like an uncomfortably big change, but @jonom and I tested it extensively (aside of the unit tests, of course), and I'll be happy to double-check any remaining compatibility concerns!

dhensby commented on filesystem/Filesystem.php:
```diff
@@ -79,6 +79,29 @@ public static function removeFolder($folder, $contentsOnly = false) {
+	 * @param Boolean $recursive Remove contained empty folders before attempting to remove this one
+	 * @return Boolean True on success, false on failure. 
+	 */
+	public static function removeFolderIfEmpty($folder, $recursive = true) {
```
static functions need to be seperated by `_`: `remove_folder_if_empty`

JorisDebonnet replied:
That confuses me a little: above and below this new method, there are existing static functions called `removeFolder`, `folderModTime`, etc?

dhensby replied:
Yep... That's SS's inconsistent naming right there. as part of 3 you could rename and deprecate the old functions so they are inline with our conventions?

dhensby commented:
Hi guys, thanks so much for this work - unfortunately this is going to need to go into 3-branch.

We need to lock down the 3.2 API. Now that 3.2.0-beta.1 has been released, we shouldn't be adding any new APIs (releasing a beta is us making an implied promise the API is stable enough that devs can rely on it and we're asking for bug reports from now on). Adding this kind of change (new API) to beta.2 would just mandate a beta.3 so that we could be sure this feature got enough time in the wild.

Ultimately, we can point at semver and say it's telling us we can't add this to 3.2.

Please can you open a new PR against 3?

JorisDebonnet commented:
Too bad this was *just* too late for beta 1, then...

So yes, I can make it against 3 instead. 

However, what about the included `base64url_encode` change in this PR, which pretty much is a bug fix? It may be rare, but `base64_encode` (whose result can now appear in resampled image filenames) can include the characters / and +. Especially the former tends to be problematic in filenames. :) 

I could fix this in a separate PR that does just that. Is the end of `Image.php` an okay place to put those helper functions?

dhensby commented:
@JorisDebonnet yes, sorry :(

Please can you patch this base64 issue against whichever is the lowest version it is in (3.1 or 3.2) then the rest against 3.

thanks a bunch, this is some great work and would have loved it in 3.2, but we should try to stick to our versioning system now

sminnee commented:
If it's any consolation, we won't need to wait nearly as long for 3.3, and if we keep the changes in the 3 branch safe, the beta/rc cycle can me much shorter.

JorisDebonnet commented:
I'll redo this against the 3 branch as soon as that branch has been brought up to date with 3.2 (since I need to build on top of the `base64url` changes from there).

dhensby commented:
@JorisDebonnet done

JorisDebonnet commented:
Redone against 3 in #4507.
